common/error.h

```diff
@@ -71,19 +73,76 @@ class [[nodiscard]] Error : public Printable<Error> {
   std::string message_;
 };
 
-// Holds a value of type `T`, or an Error explaining why the value is
+// A common base class that custom error types should derive from.
+//
+// This combines the ability to be printed with the ability to convert the error
+// to a string and in turn to a non-customized `Error` type by rendering into a
+// string.
+//
+// The goal is that custom error types can be used for errors that are common
+// and/or would have cost to fully render the error message to a string. A
+// custom type can then be used to allow custom, light-weight handling of errors
+// when appropriate. But to avoid these custom types being excessively viral, we
+// ensure they can be converted to normal `Error` types when needed by rendering
+// fully to a string.
+template <typename ErrorT>
+class [[nodiscard]] ErrorBase : public Printable<ErrorT> {
+ public:
+  // NOLINTNEXTLINE(bugprone-crtp-constructor-accessibility): Deleted.
+  ErrorBase(const ErrorBase&) = delete;
+
+  auto ToString() const -> std::string {
+    RawStringOstream s;
+    static_cast<const ErrorT*>(this)->Print(s);
+    return s.TakeStr();
+  }
+
+  explicit operator Error() { return Error(this->ToString()); }
+
+ private:
+  friend ErrorT;
```

Comment:
Why does this use `private`/`friend` instead of `protected`? Can you add a comment?

Reply:
This was the pattern that the latest clang-tidy pushed me into. It complained about protected and suggested the private + friend. I don't have a strong opinion here...

Reply:
I wouldn't have expected clang-tidy to warn on `protected`, and I'm trying to figure out which tidy warning it would be to better understand what's going on. I tried reproducing locally (with clang 20) but I don't see anything, can you share the warning you you?

Maybe add a TODO to see if we can switch to a more typical `protected` instead of friending a child class? like:

```suggestion
 // TODO: Using private/friend due to <tidy warning>, investigate switching to
 // `protected`.
 private:
  friend ErrorT;
```

Reply:
Huh... Weirdly, this doesn't reproduce for me any more. It must be contingent on some specific other bit of code.

Anyways, back to protected now.